File: Session11/C++/ZigzagLevelOrderTraversal.cpp
https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        
        if(!root)
        {
            return {};
        }
        
        vector<vector<int>> res;
        queue<TreeNode*> q;
        q.push(root);
        bool dirLtoR = true;
        while(!q.empty())
        {
            int n = q.size();
            vector<int> curr(n);
            for(int i=0; i<n; i++)
            {
                TreeNode* f = q.front();
                q.pop();
                int index = (dirLtoR) ? i : (n-1-i);
                curr[index] = f->val;
                if(f->left)
                {
                    q.push(f->left);
                }
                if(f->right)
                {
                    q.push(f->right);
                }
            }
            dirLtoR = !dirLtoR;
            res.push_back(curr);
        }
        
        return res;
        
    }
};

File: Session11/Java/ZigzagLevelOrderTraversal.java
https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        
        if(root == null)
        {
            return res;
        }
        
        queue.add(root);
        boolean dirLtoR = true; 
        while(!queue.isEmpty())
        {
            int n = queue.size();
            List<Integer> curr = new ArrayList<Integer>(n);
            for(int i=0; i<n; i++) 
            {
                TreeNode f = queue.poll();
                curr.add(f.val);
                if(f.left!=null)
                {
                    queue.add(f.left);
                }
                if(f.right!=null)
                {
                    queue.add(f.right);
                }
            }
            if(dirLtoR)
            {
                res.add(curr);
            }
            else
            {
                Collections.reverse(curr);
                res.add(curr);
            }
            dirLtoR = !dirLtoR; 
            
        }
        return res;
        
    }
}